### ISSUE TITLE
Update `MemAdviceProvider::merge_roots()` implementation to be compliant with specs

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/miden-vm#943]
Currently, `MemAdviceProvider::merge_roots()` is infallible, but according to the specs it should fail if any of the roots is not present in the Merkle store (see https://github.com/0xPolygonMiden/miden-vm/pull/933#discussion_r1231094566 for original comment).

We should update it to be compliant with specs. I believe making it compliant with specs will break MMR implementation in `stdlib`. So, we should update the implementation too to make sure it works.

**Comment by Overcastan** [comment]
Solution of this issue turned out to be quite complicated. 
MMR is the only structure which is not quite a tree, and functions which we have to manipulate the Merkle tree are not always perfect for MMR. 

Implementation of the additional check in the `merge_roots()` allow us to be sure that the nodes which we want to merge are indeed roots, and this check is working for the SMT: there cannot be a situation when we want to merge a root and an arbitrary value. At the same time for the MMR it is a common situation: every add of the node to the MMR (MMR with even number of leaves) contains a merge of two nodes, which is done using `merge_roots()`, but neither of these nodes are roots.
The crutial thing is that every node with `depth != 0` should be added to the `MerkleStore` (to get this node from the MMR we use `mtree_get`), but the only way to do so is to use the `mtree_merge`. That's why in my mind there are two options: make merging nodes a roots (by somehow adding them to the Merkle store) or merge them regardless of whether are these nodes are roots (just ignore the check).

So I have two solutions in mind:
- Create new function (instruction, decorator) to be able to add a single node as a tree root to the Merkle store (maybe in a form like `Node {parent: hash, left: zero_hash, right: zero_hash}`).
- Create essentially a copy of `merge_root` function but only for MMR without check of the roots. 
Both solutions are bad (maybe the first is a little bit better, but still).

For me the first option looks slightly better: I think I can create a decorator, which can add a node to the Merkle set. 
@bobbinth What do you think?

**Comment by bobbinth** [comment]
It's been a while since I looked at the MASM code of the MMR, but I thought all trees of the MMR with depth >= 1 are in the Merkle store. The only problem was the single leaf that was added to the MMR most recently when the number of leaves is odd. Such a leaf cannot be in the Merkle store because it has no children.

So, I thought that the problem arises only when we add a new leaf to the MMR and this leaf is odd. If that's the case, one solution could be to use an empty tree of depth 1 (where both children are `[ZERO; 4]`) and then insert relevant leaves into it (relevant nodes here would be the odd leaf and the new leaf).

This would work, but might be not super efficient (e.g., probably adds a 50 - 60 cycles).

Another solution could be as you've suggested to create another version of `mtree_merge` but which allows merging of nodes which are not in the store.

Also, we could just update the definition (and maybe rename) the `mtree_merge` operation to make this behavior explicitly allowed.

The last one is probably the simplest solution. The approach with adding a new operation may be cleaner, but I'm actually not sure about this.
